### PR TITLE
CDAP-7413 Transactional Messaging System Documentation

### PR DESCRIPTION
**Pull request by awholegunch**
Running as a Quick Build: http://builds.cask.co/browse/CDAP-DQB271-4

Page of Interest: http://builds.cask.co/artifact/CDAP-DQB271/shared/build-4/Docs-HTML/4.1.0-SNAPSHOT/en/developers-manual/building-blocks/transactional-messaging-system.html

Running as a Full Build (to test Javadocs links): http://builds.cask.co/browse/CDAP-DBT8-1 (fails, but fixed in the release/4.1 branch)

Page of Interest: http://builds.cask.co/artifact/CDAP-DBT8/shared/build-1/Current-Docs-HTML/4.1.0-SNAPSHOT/en/developers-manual/building-blocks/transactional-messaging-system.html#java-api

### CODE REVIEW

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+described below.
+
+TMS is similar to `Apache Kafka <https://kafka.apache.org>`__ but with added guarantees as
+to ordering and persistence of messages.
+
+
+Topics
+======
+Topics are created in TMS using both a namespace and a topic name. Note that while it is
+possible to fetch from ``system`` namespace, attempting to publish to the ``system``
```
Comment (gokulavasan) [inline cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: @chtyim Looks like we allow creation, deletion of topics in system namespace through MessagingAdmin. Should we disallow that?

**Comment by gokulavasan** [comment]
Text looks good. I am thinking about the illustrations. Why do we need two Workflows? And in Workflow1, why do we need Program1.1 and Program 1.2? Can't it simply be Program1 and Program2, where Program2 is watching for messages on a TMS topic and the message is visible if and only if both the message publishing and dataset updates succeeded in Program1? Or is it more complex than that? CC @chtyim

**Comment by awholegunch** [comment]
I guess it could be. @chtyim and I had started by sketching a flow, and then showing a couple of flowlets. But that was not a good example of non-transactional, because of the implied transactions in the flow. So, it became two generic programs hooked together.... but we could just have program 1 and program 2....

**Comment by gokulavasan** [comment]
@awholegunch The illustrations should revolve around a practical use-case. We can use the following:

We have two programs -> a Service and a Worker. 
A user can upload some data to the service handler and this can create a new partition and the service will also emit a message to a TMS topic.

The Worker will be polling that TMS topic and when it gets a message from it, it is guaranteed to find a new partition to work with.

**Comment by awholegunch** [comment]
Running as a Quick Build: http://builds.cask.co/browse/CDAP-DQB271-5

when finished...
Page of Interest: http://builds.cask.co/artifact/CDAP-DQB271/shared/build-5/Docs-HTML/4.1.0-SNAPSHOT/en/developers-manual/building-blocks/transactional-messaging-system.html

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+Overview
+========
+The Transactional Messaging System (TMS) is a CDAP service that provides a
+"publish-and-subscribe" messaging system that understands transactions.
+
+It uses HBase for the persistent and durable storage of messages, and supports both
+transactional and non-transactional message publishing and consumption.
+
+TMS provides ACID *(atomicity, consistency, isolation,* and *durability)* guarantees,
+using `Apache Tephra <http://tephra.incubator.apache.org>`__ for the ACI portion. TMS
```
Comment (chtyim) [inline cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: No need to call out `ACI`. In fact the TMS itself has nothing to do with ACID (probably just the durable part). Probably better just not to have this paragraph.

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+doesn't interact directly with Tephra; instead, the client provides the transaction, as
+described below.
+
+TMS is similar to `Apache Kafka <https://kafka.apache.org>`__ but with added guarantees as
```
Comment (chtyim) [inline cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: Better not to mention Kafka in here as we don't want users to use TMS for transporting actual user data (which Kafka does), as the volume can be very high and TMS is not yet optimized for those use cases.

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+
+Message Publishing
+==================
+Messages in TMS are published as either `non-transactional <Non-transactional Messages>`__
+or `transactional <Transactional Messages>`__ messages.
+
+Non-transactional Message Publishing
+------------------------------------
+Non-transactional messaging in TMS works in a manner very similar to `Apache Kafka
```
Comment (chtyim) [inline cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: Again, better not to mention Kafka.

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+TMS provides strong ordering guarantees for the consumers of a topic:
+
+- When fetching from the same topic, every consumer will see the exact same order of
+  messages.
+
+- A consumer, fetching from the same topic more than once from the same point, will always
+  see the topics in the exact same order.
+
+Each message has a timestamp, which can be thought of as the published time, or the time
+the system persisted the message. Each message ID, combined with a topic, uniquely
```
Comment (chtyim) [inline cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: The `message ID` seems out of context. It just suddenly popped up without explanation.

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+Messages in TMS are published as either `non-transactional <Non-transactional Messages>`__
+or `transactional <Transactional Messages>`__ messages.
+
+Non-transactional Message Publishing
+------------------------------------
+Non-transactional messaging in TMS works in a manner very similar to `Apache Kafka
+<https://kafka.apache.org>`__: 
+
+- **publish** a message (also called a *payload*) to a topic
+- **fetch** (also called *subscribe* or *consume*) from a topic
```
Comment (chtyim) [inline cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: I suppose this section is talking about publishing, shall we talk about consumption later??

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+      while (!iterator.hasNext() && stopwatch.elapsedTime(unit) < timeout) {
+        TimeUnit.MILLISECONDS.sleep(100);
+        iterator = fetcher.fetch(namespace, topic, 1, afterMessageId);
+      }
+
+      if (!iterator.hasNext()) {
+        throw new TimeoutException("Failed to get any messages from " + topic +
+                                     " in " + timeout + " " + unit.name().toLowerCase());
+      }
+      // The payload contains the message to publish in next step
```
Comment (chtyim) [inline cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: Remove this comment.

**Comment by chtyim** [comment]
The Java code example is not good enough. Need to have better, cleaner examples/code fragment.

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst (hunk 2 above)
Comment (awholegunch) [reply cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: Fixed.

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst (hunk 3 above)
Comment (awholegunch) [reply cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: Fixed.

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst (hunk 4 above)
Comment (awholegunch) [reply cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: Fixed.

Comment (awholegunch) [reply cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: Fixed.

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst (hunk 7 above)
Comment (awholegunch) [reply cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: Fixed.

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst (hunk 5 above)
Comment (awholegunch) [reply cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: Fixed.

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst (hunk 6 above)
Comment (awholegunch) [reply cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: Re-named sections as they talk about both publishing and consumption. However, is there more we _need to_ or _can_ say about consumption?

**Comment by awholegunch** [comment]
Running as a Quick Build: http://builds.cask.co/browse/CDAP-DQB271-6

Page of Interest: http://builds.cask.co/artifact/CDAP-DQB271/shared/build-6/Docs-HTML/4.1.0-SNAPSHOT/en/developers-manual/building-blocks/transactional-messaging-system.html

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+    :copyright: Copyright © 2017 Cask Data, Inc.
+    :description: Transactional Messaging System, a ACID-guaranteed "publish-and-subscribe" messaging service
+
+.. _transactional-messaging-system:
+
+==============================
+Transactional Messaging System
+==============================
+
+.. topic::  **Note: Beta Feature**
```
Comment (sreevatsanraman) [inline cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: We can remove this. We already say this below.

**Comment by sreevatsanraman** [comment]
One comment. Open a JIRA for improving examples. LGTM otherwise.

File: cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst (hunk 8 above)
Comment (awholegunch) [reply cdap-docs/developers-manual/source/building-blocks/transactional-messaging-system.rst]: Removed duplicate "beta".

**Comment by awholegunch** [comment]
Jira for code samples: https://issues.cask.co/browse/CDAP-8824

**Pull request merged**